File: Linear Shoal/Stack/Q2. Evaluate Reverse Polish Notation/Q2_EvaluateReversePolishNotation.java
import java.util.Stack;

class Solution {
    public int evalRPN(String[] tokens) {
        // Stack to store numbers during evaluation
        Stack<Integer> stack = new Stack<>();
        
        // Loop through each token in the input
        for (String token : tokens) {
            // If the token is an operator
            if (token.equals("+") || token.equals("-") || token.equals("*") || token.equals("/")) {
                // Pop the second operand from stack
                int b = stack.pop();
                // Pop the first operand from stack
                int a = stack.pop();
                // Variable to store result of operation
                int res = 0;
                
                // Apply the correct operation
                switch (token) {
                    case "+": res = a + b; break; // addition
                    case "-": res = a - b; break; // subtraction
                    case "*": res = a * b; break; // multiplication
                    case "/": res = a / b; break; // division (truncates toward zero in Java)
                }
                // Push the result back onto the stack
                stack.push(res);
            } else {
                // If token is a number, convert string to integer and push onto stack
                stack.push(Integer.parseInt(token));
            }
        }
        
        // At the end, stack has one element: the final result
        return stack.pop();
    }
}

File: Linear Shoal/Stack/Q2. Evaluate Reverse Polish Notation/Q2_EvaluateReversePolishNotation.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []  # stack to hold numbers during evaluation
        
        for token in tokens:
            if token in {"+", "-", "*", "/"}:
                # Pop the last two numbers
                b = stack.pop()
                a = stack.pop()
                
                # Perform the operation
                if token == "+":
                    stack.append(a + b)
                elif token == "-":
                    stack.append(a - b)
                elif token == "*":
                    stack.append(a * b)
                elif token == "/":
                    # Division should truncate toward zero
                    stack.append(int(a / float(b)))
            else:
                # If token is a number, convert and push to stack
                stack.append(int(token))
        
        # Final result will be the only number left in stack
        return stack.pop()
